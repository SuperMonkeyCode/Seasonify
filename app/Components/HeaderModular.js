/* @flow */
import React from 'react';
import { Keyboard } from 'react-native';
import { Header, Title, Left, Right, Body, Button, Icon, Text } from 'native-base';
import type { HeaderButton } from 'Types';

export default function HeaderModular({
  title,
  cancelButton,
  actionButtons = [],
}: {
  title: string,
  cancelButton: HeaderButton,
  actionButtons: HeaderButton[],
}) {
  const cancelAction = () => {
    cancelButton.action();
    Keyboard.dismiss();
  };

  const renderTextOrIcon = (button: HeaderButton) => {
    if (button.icon) return <Icon name={button.icon} />;
    if (button.text) return <Text>{button.text}</Text>;
    return null;
  };

  const renderButton = () =>
    actionButtons.map((button, index) => {
      if (
        button.visibleIf ||
        (typeof button.visibleIf === 'undefined' && !button.hideByDefault)
      ) {
        return (
          /* eslint-disable react/no-array-index-key */
          <Button key={index} transparent onPress={button.action}>
            {renderTextOrIcon(button)}
          </Button>
        );
      }
      return null;
    });

  return (
    <Header>
      <Left>
        <Button transparent onPress={cancelAction}>
          {renderTextOrIcon(cancelButton)}
        </Button>
      </Left>
      <Body>
        <Title>{title}</Title>
      </Body>
      <Right>
        {renderButton()}
      </Right>
    </Header>
  );
}
