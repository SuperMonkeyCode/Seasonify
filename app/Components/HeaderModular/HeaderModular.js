/* @flow */
import { Header, Title, Left, Right, Body, Button, Icon, Text } from 'native-base';
import { Keyboard } from 'react-native';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';

import type { HeaderButton } from 'Types';

type Props = {
  /* parent */
  title: string,
  cancelButton: HeaderButton,
  actionButtons?: HeaderButton[],
  /* connect */
  /* HOC */
  handleCancel: Function,
};

const enhance = compose(
  pure,
  withHandlers({
    handleCancel: ({ cancelButton }: Props) => () => {
      cancelButton.action();
      Keyboard.dismiss();
    },
  }),
);

function HeaderModular({ title, cancelButton, actionButtons = [], handleCancel }: Props) {
  const renderTextOrIcon = (button: HeaderButton) => {
    if (button.icon) return <Icon name={button.icon} />;
    if (button.text) {
      return (
        <Text>
          {button.text}
        </Text>
      );
    }
    return null;
  };

  const renderButton = () =>
    actionButtons.map((button, index) => {
      if (
        button.visibleIf ||
        (typeof button.visibleIf === 'undefined' && !button.hideByDefault)
      ) {
        return (
          /* eslint-disable react/no-array-index-key */
          <Button key={index} transparent onPress={button.action}>
            {renderTextOrIcon(button)}
          </Button>
        );
      }
      return null;
    });

  return (
    <Header>
      <Left>
        <Button transparent onPress={handleCancel}>
          {renderTextOrIcon(cancelButton)}
        </Button>
      </Left>
      <Body>
        <Title>
          {title}
        </Title>
      </Body>
      <Right>
        {renderButton()}
      </Right>
    </Header>
  );
}

export default enhance(HeaderModular);
