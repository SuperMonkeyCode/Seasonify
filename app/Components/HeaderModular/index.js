import HeaderModular from './HeaderModular';

export default HeaderModular;
