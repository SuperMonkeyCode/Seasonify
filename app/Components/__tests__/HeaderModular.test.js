import React from 'react';

import HeaderModular from '../HeaderModular';

function setup(specificProps = {}) {
  const props = {
    title: 'Some title',
    cancelButton: { icon: 'icon', action: jest.fn() },
    ...specificProps,
  };
  const component = shallow(<HeaderModular {...props} />);
  return {
    component,
    props,
  };
}

const actionButtons = [
  {
    visibleIf: true,
    icon: 'icon1',
    action: jest.fn(),
  },
  {
    visibleIf: false,
    text: 'icon2',
    action: jest.fn(),
  },
  {
    visibleIf: true,
    hideByDefault: true,
    icon: 'icon3',
    action: jest.fn(),
  },
  {
    visibleIf: true,
    hideByDefault: false,
    icon: 'icon4',
    action: jest.fn(),
  },
  { text: 'text1', action: jest.fn() },
  { action: jest.fn() },
];

describe('Rendering', () => {
  it('should render with a basic cancel button', () => {
    const { component } = setup();
    expect(component).toMatchSnapshot();
  });

  it('should render all visible action buttons', () => {
    const { component } = setup({ actionButtons });
    const buttons = component.find('Styled(Button)');
    expect(buttons).toHaveLength(6);
    expect(component).toMatchSnapshot();
  });
});

describe('Events', () => {
  it('should call cancelButton.action', () => {
    const { component, props } = setup();
    component.find('Styled(Button)').simulate('press');
    expect(props.cancelButton.action).toBeCalled();
  });

  it('should call last actionButtons action', () => {
    const { component, props } = setup({ actionButtons });
    component.find('Styled(Button)').last().simulate('press');
    expect(props.actionButtons[actionButtons.length - 1].action).toBeCalled();
  });
});
