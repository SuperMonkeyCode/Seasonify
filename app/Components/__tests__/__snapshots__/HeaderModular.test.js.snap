// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering should render all visible action buttons 1`] = `
<Styled(Header)>
  <Styled(Left)>
    <Styled(Button)
      onPress={[Function]}
      transparent={true}
    >
      <Styled(Icon)
        name="icon"
      />
    </Styled(Button)>
  </Styled(Left)>
  <Styled(Body)>
    <Styled(Title)>
      Some title
    </Styled(Title)>
  </Styled(Body)>
  <Styled(Right)>
    <Styled(Button)
      onPress={[Function]}
      transparent={true}
    >
      <Styled(Icon)
        name="icon1"
      />
    </Styled(Button)>
    <Styled(Button)
      onPress={[Function]}
      transparent={true}
    >
      <Styled(Icon)
        name="icon3"
      />
    </Styled(Button)>
    <Styled(Button)
      onPress={[Function]}
      transparent={true}
    >
      <Styled(Icon)
        name="icon4"
      />
    </Styled(Button)>
    <Styled(Button)
      onPress={[Function]}
      transparent={true}
    >
      <Styled(Text)>
        text1
      </Styled(Text)>
    </Styled(Button)>
  </Styled(Right)>
</Styled(Header)>
`;

exports[`Rendering should render with a basic cancel button 1`] = `
<Styled(Header)>
  <Styled(Left)>
    <Styled(Button)
      onPress={[Function]}
      transparent={true}
    >
      <Styled(Icon)
        name="icon"
      />
    </Styled(Button)>
  </Styled(Left)>
  <Styled(Body)>
    <Styled(Title)>
      Some title
    </Styled(Title)>
  </Styled(Body)>
  <Styled(Right) />
</Styled(Header)>
`;
